File: scripts/calculateStandings.ts
import { pointsPenalties, results, seasonRacers } from "@/data";
import type { SeasonName, TrackName, RacerName } from "@/types";
import { calcRacePoints } from "@/racePoints";
import { doublePointsTracks } from "@/points";

type BaseStanding = {
  points: number;
  netPoints: number;
  delta: number;
  netDelta: number;
};

type DriverStanding = BaseStanding & {
  deductedRaces: string[];
};

type ConstructorStanding = BaseStanding & {
  normalisedPoints: number;
  netNormalisedPoints: number;
  driverCount: number;
};

type Standings = Record<
  string,
  Record<string, DriverStanding | ConstructorStanding>
>;

class StandingsCalculator {
  private readonly standardDriverCount = 2;
  private points: Record<string, Record<string, number>> = {};
  private worstResults: Record<
    string,
    Record<string, { races: string[]; points: number }>
  > = {};

  constructor(
    private season: SeasonName,
    private races: string[],
  ) {}

  private getDriverCar(driver: RacerName, race: TrackName): string {
    const seasonRacer = seasonRacers[this.season][driver];
    return (
      seasonRacer?.otherTeams?.[race]?.car ?? seasonRacer?.car ?? "unknown"
    );
  }

  private calculateRacePoints(race: TrackName): Record<string, number> {
    const raceResults = results[this.season][race];
    if (!raceResults?.results) return {};

    const classOf = (id: string) =>
      seasonRacers[this.season][id as RacerName]?.class;

    // Multiclass seasons store per-class fastest laps; single-class seasons fall
    // back to the one overall fastest lap (bucketed under its driver's class or "all").
    const fl = raceResults.fastestLap;
    const classFastestLaps =
      raceResults.classFastestLaps ??
      (fl ? { [classOf(fl.racerId) ?? "all"]: fl.racerId } : {});

    const multiplier = (doublePointsTracks[this.season] ?? []).includes(race)
      ? 2
      : 1;

    return calcRacePoints({
      season: this.season,
      orderedResults: Object.entries(raceResults.results) as [string, string][],
      classOf,
      classFastestLaps,
      penalties: pointsPenalties[this.season][race] as
        | Record<string, number>
        | undefined,
      multiplier,
    });
  }

  private calculateWorstResults() {
    // Track worst results per driver up to each race
    this.races.forEach((race, raceIndex) => {
      this.worstResults[race] = {};

      Object.keys(seasonRacers[this.season] ?? {}).forEach((driver) => {
        // Get all race results up to this point
        const raceResults = this.races
          .slice(0, raceIndex + 1)
          .map((raceKey) => ({
            race: raceKey,
            // Get the non-cumulative points for this specific race
            points:
              (this.calculateRacePoints(raceKey as TrackName) ?? {})[driver] ??
              0,
          }))
          .sort((a, b) => a.points - b.points); // Sort ascending to get worst results first

        // Take the two worst results
        const worstTwo = raceResults.slice(0, 2);

        this.worstResults[race][driver] = {
          races: worstTwo.map((r) => r.race),
          points: worstTwo.reduce((sum, r) => sum + r.points, 0),
        };
      });
    });
  }

  private getAllSeasonDrivers(): string[] {
    return Object.keys(seasonRacers[this.season] ?? {});
  }

  private fillMissingDriverPoints(race: string, index: number) {
    const allDrivers = this.getAllSeasonDrivers();

    allDrivers.forEach((driver) => {
      // If driver didn't race, set their points to 0 for this race
      if (!this.points[race][driver]) {
        this.points[race][driver] = 0;
      }
    });
  }

  private calculateDriverStandings(): Record<
    string,
    Record<string, DriverStanding>
  > {
    const standings: Record<string, Record<string, DriverStanding>> = {};
    const allDrivers = this.getAllSeasonDrivers();

    this.races.forEach((race, index) => {
      if (!this.points[race]) return;

      standings[race] = {};
      const currentPoints = this.points[race];
      const previousStandings =
        index > 0 ? standings[this.races[index - 1]] : null;

      allDrivers.forEach((driver) => {
        const points = currentPoints[driver] ?? 0;
        // Subtract the worst two results for this driver up to this race
        const netPoints =
          points - (this.worstResults[race][driver]?.points ?? 0);

        standings[race][driver] = {
          points,
          netPoints,
          delta: 0,
          netDelta: 0,
          deductedRaces: this.worstResults[race][driver]?.races ?? [],
        };
      });

      this.updatePositionDeltas(standings[race], previousStandings);
    });

    return standings;
  }

  private calculateConstructorStandings(): Record<
    string,
    Record<string, ConstructorStanding>
  > {
    const standings: Record<string, Record<string, ConstructorStanding>> = {};

    this.races.forEach((race, index) => {
      if (!this.points[race]) return;

      standings[race] = {};
      const previousStandings =
        index > 0 ? standings[this.races[index - 1]] : null;
      const raceResults =
        results[this.season][race as TrackName]?.results ?? {};

      // Get all constructors
      const allConstructors = new Set<string>();
      this.getAllSeasonDrivers().forEach((driver) => {
        allConstructors.add(
          this.getDriverCar(driver as RacerName, race as TrackName),
        );
      });

      // Calculate points for this specific race
      const raceConstructorPoints: Record<
        string,
        {
          racePoints: number;
          driverCount: number;
          drivers: string[];
        }
      > = {};

      // Count participating drivers and calculate points
      Object.keys(raceResults).forEach((driver) => {
        const car = this.getDriverCar(driver as RacerName, race as TrackName);
        if (!raceConstructorPoints[car]) {
          raceConstructorPoints[car] = {
            racePoints: 0,
            driverCount: 0,
            drivers: [],
          };
        }
        raceConstructorPoints[car].driverCount += 1;
        raceConstructorPoints[car].drivers.push(driver);
      });

      // Add points for scoring finishes
      const singleRacePoints =
        this.calculateRacePoints(race as TrackName) ?? {};
      Object.entries(singleRacePoints).forEach(([driver, points]) => {
        const car = this.getDriverCar(driver as RacerName, race as TrackName);
        raceConstructorPoints[car].racePoints += points;
      });

      // Calculate standings for each constructor
      allConstructors.forEach((car) => {
        const currentRaceData = raceConstructorPoints[car];

        if (!currentRaceData || currentRaceData.driverCount === 0) {
          // Constructor didn't participate, carry forward previous standings
          if (previousStandings?.[car]) {
            standings[race][car] = { ...previousStandings[car] };
          }
          return;
        }

        // Calculate normalized points for this race
        const normalizationFactor =
          this.standardDriverCount / currentRaceData.driverCount;
        const raceNormalizedPoints = Math.round(
          currentRaceData.racePoints * normalizationFactor,
        );

        // Get previous standings or start at 0
        const previousPoints = previousStandings?.[car]?.points ?? 0;
        const previousNormalizedPoints =
          previousStandings?.[car]?.normalisedPoints ?? 0;

        // Calculate cumulative totals
        const totalPoints = previousPoints + currentRaceData.racePoints;
        const totalNormalizedPoints =
          previousNormalizedPoints + raceNormalizedPoints;

        // Find two worst results up to this point, including non-participation (0 points)
        const constructorResults = this.races
          .slice(0, index + 1)
          .map((raceKey) => {
            const raceData = results[this.season][raceKey as TrackName];

            // If no results or no participation, count as 0 points
            if (!raceData?.results)
              return { points: 0, normalized: 0, race: raceKey };

            const drivers = Object.keys(raceData.results).filter(
              (d) =>
                this.getDriverCar(d as RacerName, raceKey as TrackName) === car,
            );

            // No participation counts as 0 points
            if (drivers.length === 0)
              return { points: 0, normalized: 0, race: raceKey };

            // Calculate points excluding drivers with penalties
            const racePoints = drivers.reduce((sum, d) => {
              // Skip this driver's points if they have a penalty
              if (
                pointsPenalties[this.season][raceKey as TrackName]?.[
                  d as RacerName
                ] !== undefined
              ) {
                return sum;
              }
              return (
                sum + (this.calculateRacePoints(raceKey as TrackName)?.[d] ?? 0)
              );
            }, 0);

            // Only count drivers without penalties for normalization
            const driversWithoutPenalties = drivers.filter(
              (d) =>
                pointsPenalties[this.season][raceKey as TrackName]?.[
                  d as RacerName
                ] === undefined,
            );

            // Calculate normalization based on drivers without penalties
            const effectiveDriverCount =
              driversWithoutPenalties.length || drivers.length;
            const normalizationFactor =
              this.standardDriverCount / effectiveDriverCount;
            const normalizedPoints = Math.round(
              racePoints * normalizationFactor,
            );

            return {
              points: racePoints,
              normalized: normalizedPoints,
              race: raceKey,
            };
          });

        // Get two worst results by normalized points
        const worstNormalizedResults = [...constructorResults]
          .sort((a, b) => a.normalized - b.normalized)
          .slice(0, 2);

        const worstResultsPoints = [...constructorResults]
          .sort((a, b) => a.points - b.points)
          .slice(0, 2)
          .reduce((sum, r) => sum + r.points, 0);

        const worstResultsNormalized = worstNormalizedResults.reduce(
          (sum, r) => sum + r.normalized,
          0,
        );

        standings[race][car] = {
          points: totalPoints,
          normalisedPoints: totalNormalizedPoints,
          netPoints: totalPoints - worstResultsPoints,
          netNormalisedPoints: totalNormalizedPoints - worstResultsNormalized,
          driverCount: currentRaceData.driverCount,
          delta: 0,
          netDelta: 0,
        };
      });

      this.updatePositionDeltas(standings[race], previousStandings);
    });

    return standings;
  }

  // Rank drivers within their class (or a single "all" bucket for single-class
  // seasons) by the given metric, returning a 1-based position per driver.
  private rankByClass(
    standings: Record<string, BaseStanding>,
    metric: "points" | "netPoints",
  ): Map<string, number> {
    const byClass: Record<string, string[]> = {};
    Object.keys(standings).forEach((id) => {
      const cls = seasonRacers[this.season][id as RacerName]?.class ?? "all";
      (byClass[cls] ??= []).push(id);
    });

    const positions = new Map<string, number>();
    Object.values(byClass).forEach((ids) => {
      ids
        .sort(
          (a, b) => (standings[b][metric] ?? 0) - (standings[a][metric] ?? 0),
        )
        .forEach((id, index) => positions.set(id, index + 1));
    });
    return positions;
  }

  private updatePositionDeltas<T extends BaseStanding | ConstructorStanding>(
    currentStandings: Record<string, T>,
    previousStandings: Record<string, T> | null,
  ) {
    // For the first race, all deltas should be 0
    if (!previousStandings) {
      Object.values(currentStandings).forEach((standing) => {
        standing.delta = 0;
        standing.netDelta = 0;
      });
      return;
    }

    if (this.isConstructorStanding(currentStandings)) {
      // TypeScript now knows this is Record<string, ConstructorStanding>
      const currentPositions = new Map(
        Object.entries(currentStandings)
          .sort(([, a], [, b]) => b.normalisedPoints - a.normalisedPoints)
          .map(([id], index) => [id, index + 1]),
      );

      const previousPositions = new Map(
        Object.entries(previousStandings as Record<string, ConstructorStanding>)
          .sort(([, a], [, b]) => b.normalisedPoints - a.normalisedPoints)
          .map(([id], index) => [id, index + 1]),
      );

      // Calculate regular delta using normalised points
      Object.keys(currentStandings).forEach((id) => {
        const currentPos = currentPositions.get(id) ?? 0;
        const previousPos = previousPositions.get(id) ?? currentPos;
        currentStandings[id].delta = previousPos - currentPos;
      });

      // Calculate net delta using net normalised points
      const currentNetPositions = new Map(
        Object.entries(currentStandings)
          .sort(([, a], [, b]) => b.netNormalisedPoints - a.netNormalisedPoints)
          .map(([id], index) => [id, index + 1]),
      );

      const previousNetPositions = new Map(
        Object.entries(previousStandings as Record<string, ConstructorStanding>)
          .sort(([, a], [, b]) => b.netNormalisedPoints - a.netNormalisedPoints)
          .map(([id], index) => [id, index + 1]),
      );

      Object.keys(currentStandings).forEach((id) => {
        const currentPos = currentNetPositions.get(id) ?? 0;
        const previousPos = previousNetPositions.get(id) ?? currentPos;
        currentStandings[id].netDelta = previousPos - currentPos;
      });
    } else {
      // Drivers are ranked within their class, so position deltas reflect
      // movement in the class championship (a leader who stays top of their
      // class shows no change, even if passed on overall points). Single-class
      // seasons collapse to one "all" bucket, i.e. overall ranking as before.
      const currentPositions = this.rankByClass(currentStandings, "points");
      const previousPositions = this.rankByClass(previousStandings, "points");

      Object.keys(currentStandings).forEach((id) => {
        const currentPos = currentPositions.get(id) ?? 0;
        const previousPos = previousPositions.get(id) ?? currentPos;
        currentStandings[id].delta = previousPos - currentPos;
      });

      // Calculate net delta using net points for drivers
      const currentNetPositions = this.rankByClass(
        currentStandings,
        "netPoints",
      );
      const previousNetPositions = this.rankByClass(
        previousStandings,
        "netPoints",
      );

      Object.keys(currentStandings).forEach((id) => {
        const currentPos = currentNetPositions.get(id) ?? 0;
        const previousPos = previousNetPositions.get(id) ?? currentPos;
        currentStandings[id].netDelta = previousPos - currentPos;
      });
    }
  }

  // Type guard to check if we're dealing with constructor standings
  private isConstructorStanding(
    standings: Record<string, BaseStanding | ConstructorStanding>,
  ): standings is Record<string, ConstructorStanding> {
    const firstStanding = Object.values(standings)[0];
    return "normalisedPoints" in firstStanding;
  }

  public calculate(): [
    Record<string, Record<string, DriverStanding>>,
    Record<string, Record<string, ConstructorStanding>>,
  ] {
    // Calculate initial race points
    this.races.forEach((race, index) => {
      this.points[race] = this.calculateRacePoints(race as TrackName) ?? {};
      this.fillMissingDriverPoints(race, index);
    });

    // Calculate cumulative points
    this.races.forEach((race, index) => {
      if (index === 0) return;

      const allDrivers = this.getAllSeasonDrivers();
      const previousRace = this.races[index - 1];

      allDrivers.forEach((driver) => {
        // Add previous race points to current race points (which may be 0 if they didn't race)
        this.points[race][driver] += this.points[previousRace][driver] ?? 0;
      });
    });

    this.calculateWorstResults();

    return [
      this.calculateDriverStandings(),
      this.calculateConstructorStandings(),
    ];
  }
}

export const calculateStandings = (season: SeasonName) => {
  const completedRaces = Object.entries(results[season])
    .filter(([, raceData]) => raceData?.results !== null)
    .map(([race]) => race);

  const calculator = new StandingsCalculator(season, completedRaces);
  return calculator.calculate();
};

File: scripts/calculate_standings.rs
use std::collections::HashMap;
use crate::data::{points_penalties, results, season_racers};
use crate::types::{SeasonName, TrackName, RacerName};
use crate::standings::pointsless_results;
use crate::data::results::RaceResults;
use crate::points::points_scheme;
use crate::data::penalties::Penalties;

type GeneratedRaceStandings = HashMap<String, i32>;
type GeneratedStandings = HashMap<String, Option<GeneratedRaceStandings>>;

const STANDARD_DRIVER_COUNT: i32 = 2;

pub fn calculate_standings(season: SeasonName) -> (HashMap<String, HashMap<String, HashMap<String, i32>>>, HashMap<String, HashMap<String, HashMap<String, i32>>>) {
    let season_races: Vec<&str> = results[season].keys().collect();
    let mut driver_points: HashMap<String, HashMap<String, HashMap<String, i32>>> = HashMap::new();
    let mut constructor_points: HashMap<String, HashMap<String, HashMap<String, i32>>> = HashMap::new();

    let points: GeneratedStandings = season_races.iter().fold(HashMap::new(), |mut races_obj, &race| {
        let race_results = results[season].get(race).unwrap_or(&None);

        let pts = if let Some(race_results) = race_results {
            race_results.results.iter().enumerate().fold(HashMap::new(), |mut obj, (current_index, (driver, result))| {
                let dnf = pointsless_results.contains(result);
                let cumulative_points = (points_scheme[season][current_index] as i32)
                    + if race_results.fastest_lap.as_ref().map_or(false, |fl| fl.racer_id == *driver) { 1 } else { 0 }
                    - points_penalties[season][race].get(driver).cloned().unwrap_or(0);

                obj.insert(driver.clone(), if dnf { 0 } else { cumulative_points });
                obj
            })
        } else {
            None
        };

        races_obj.insert(race.to_string(), pts);
        races_obj
    });

    let race_keys: Vec<&str> = points.keys().map(|k| k.as_str()).collect();

    for (index, &race) in race_keys.iter().enumerate() {
        if points[race].is_none() {
            continue;
        }

        let race_points = points[race].as_ref().unwrap();
        let previous_race_points = if index > 0 { points[race_keys[index - 1]].as_ref().unwrap() } else { &HashMap::new() };
        let race_points_keys: Vec<&str> = race_points.keys().map(|k| k.as_str()).collect();

        for driver in race_points_keys.iter() {
            let season_racer = season_racers[season].get(*driver).unwrap_or(&None);
            let car = season_racer.and_then(|sr| sr.other_teams.get(race)).map_or_else(|| season_racer.map_or("unknown", |sr| sr.car.as_str()), |ot| ot.car.as_str());
            let added_constructor_points = constructor_points.get(race).and_then(|cp| cp.get(car)).map_or(0, |cp| cp.points) + race_points[*driver];

            constructor_points.entry(race.to_string()).or_insert_with(HashMap::new).entry(car.to_string()).or_insert_with(|| HashMap::new()).insert("points".to_string(), added_constructor_points);
            constructor_points.entry(race.to_string()).or_insert_with(HashMap::new).entry(car.to_string()).or_insert_with(|| HashMap::new()).insert("normalised_points".to_string(), added_constructor_points);
            constructor_points.entry(race.to_string()).or_insert_with(HashMap::new).entry(car.to_string()).or_insert_with(|| HashMap::new()).insert("driver_count".to_string(), constructor_points.get(race).and_then(|cp| cp.get(car)).map_or(0, |cp| cp.driver_count) + 1);
            constructor_points.entry(race.to_string()).or_insert_with(HashMap::new).entry(car.to_string()).or_insert_with(|| HashMap::new()).insert("delta".to_string(), 0);

            if index > 0 {
                race_points.insert(driver.to_string(), race_points[driver] + previous_race_points.get(driver).cloned().unwrap_or(0));
            }
        }

        if index > 0 {
            for driver in season_racers[season].keys() {
                if !race_points_keys.contains(&driver.as_str()) {
                    race_points.insert(driver.to_string(), previous_race_points.get(driver).cloned().unwrap_or(0));
                }
            }
        }

        if let Some(constructor_points_race) = constructor_points.get_mut(race) {
            for (constructor, points) in constructor_points_race.iter_mut() {
                points.insert("normalised_points".to_string(), (points["points"] as f32 * (STANDARD_DRIVER_COUNT as f32 / points["driver_count"] as f32)).round() as i32);
            }

            let cumulative_points: Vec<_> = constructor_points_race.iter_mut().map(|(constructor, points)| {
                if index > 0 {
                    points.insert("points".to_string(), points["points"] + constructor_points[race_keys[index - 1]].get(constructor).map_or(0, |cp| cp["points"]));
                    points.insert("normalised_points".to_string(), points["normalised_points"] + constructor_points[race_keys[index - 1]].get(constructor).map_or(0, |cp| cp["normalised_points"]));
                }
                (constructor.clone(), points.clone())
            }).collect();

            if index > 0 {
                let missed: Vec<_> = constructor_points[race_keys[index - 1]].keys().filter(|constructor| !cumulative_points.iter().any(|(c, _)| c == *constructor)).cloned().collect();
                for constructor in missed {
                    cumulative_points.push((constructor.clone(), constructor_points[race_keys[index - 1]].get(&constructor).unwrap().clone()));
                }
            }

            *constructor_points_race = cumulative_points.into_iter().sorted_by(|(_, a), (_, b)| b["normalised_points"].cmp(&a["normalised_points"])).collect();
        }

        *points.get_mut(race).unwrap() = race_points.iter().sorted_by(|(_, a), (_, b)| b.cmp(a)).map(|(k, v)| (k.clone(), *v)).collect();

        let driver_keys: Vec<&str> = points[race].keys().map(|k| k.as_str()).collect();
        let previous_driver_keys: Vec<&str> = if index > 0 { points[race_keys[index - 1]].keys().map(|k| k.as_str()).collect() } else { driver_keys.clone() };

        for driver in driver_keys.iter() {
            let delta = previous_driver_keys.iter().position(|&d| d == *driver).unwrap_or(0) as i32 - driver_keys.iter().position(|&d| d == *driver).unwrap_or(0) as i32;

            driver_points.entry(race.to_string()).or_insert_with(HashMap::new).entry(driver.to_string()).or_insert_with(|| HashMap::new()).insert("points".to_string(), points[race][*driver]);
            driver_points.entry(race.to_string()).or_insert_with(HashMap::new).entry(driver.to_string()).or_insert_with(|| HashMap::new()).insert("delta".to_string(), if index > 0 { delta } else { 0 });
        }

        let constructor_keys: Vec<&str> = constructor_points[race].keys().map(|k| k.as_str()).collect();
        let constructor_previous_keys: Vec<&str> = if index > 0 { constructor_points[race_keys[index - 1]].keys().map(|k| k.as_str()).collect() } else { constructor_keys.clone() };

        for constructor in constructor_keys.iter() {
            let delta = constructor_previous_keys.iter().position(|&c| c == *constructor).unwrap_or(0) as i32 - constructor_keys.iter().position(|&c| c == *constructor).unwrap_or(0) as i32;

            constructor_points.entry(race.to_string()).or_insert_with(HashMap::new).entry(constructor.to_string()).or_insert_with(|| HashMap::new()).insert("delta".to_string(), if index > 0 { delta } else { 0 });
        }
    }

    (driver_points, constructor_points)
}
